### PR TITLE
improve CA_NS validation and logging

### PR DESCRIPTION
**Pull request by jarek**
Hello,

Here are a few improvements to Nova Scotia:

* We are now following the validation logic used on their website (https://www.nspower.ca/site/renewables/assets/js/site.js), plus checking that reported generation doesn't go greatly over known capacity. The JSON data is sometimes just plain wrong, and doesn't correct afterward.
* Exchange parser now returns all available data rather than only the latest point.
* We now use logger. See question attached below about whether it should be warning or info.

Screenshot of problem:

ns

### CODE REVIEW

File: parsers/CA_NS.py [hunk 1]
```diff
@@ -18,35 +44,73 @@ def _get_ns_info(requests_obj):
...
+        # validate again
+        valid = True
+        for gen_type, value in electricity_mix.items():
+            absolute_bound = valid_absolute.get(gen_type)  # imports are not in valid_absolute
+            if absolute_bound and value > absolute_bound:
+                valid = False
+                logger.warning(
+                    'CA-NS: discarding datapoint at {dt} '
+                    'due to {fuel} too high: {value}'.format(
+                        dt=data_date, fuel=gen_type, value=value))
```
Comment (jarek) [inline parsers/CA_NS.py]: This isn't a fatal problem, it happens every now and then due to data source problems, and we recover from it gracefully. Should we use logger.warning or logger.info?

File: parsers/CA_NS.py [hunk 2]
```diff
@@ -18,35 +44,73 @@ def _get_ns_info(requests_obj):
...
             # if not found, assume 1244 MW, based on average yearly electricity available for use
             # in 2014 and 2015 (Statistics Canada table Table 127-0008 for Nova Scotia)
-            load = 1244 / 100.0
-
-        # datetime is in format '/Date(1493924400000)/'
-        # get the timestamp 1493924400 (cutting out last three zeros as well)
-        data_timestamp = int(mix['datetime'][6:-5])
-        data_date = arrow.get(data_timestamp).datetime
+            load = 1244
+            logger.warning('CA-NS: unable to find load for {}, assuming 1244 MW'.format(data_date))
```
Comment (corradio) [inline parsers/CA_NS.py]: can you add `extra={'key': 'CA-NS'}` as extra argument to make sure the zone code gets registered?

File: parsers/CA_NS.py (hunk 1 above)
Comment (corradio) [reply parsers/CA_NS.py]: Warning seems like the proper level to me, as you're discarding data.

File: parsers/CA_NS.py (hunk 2 above)
Comment (jarek) [reply parsers/CA_NS.py]: Sure - how should it look in total? Is there an existing example? Like this?

```
logger.warning(
    'CA-NS: unable to find load for {}, assuming 1244 MW'.format(data_date),
    extra={'key':'CA-NS'})
```

And do that in all warnings I log?

File: parsers/CA_NS.py (hunk 1 above)
Comment (jarek) [reply parsers/CA_NS.py]: Yeah, on second thought, everything that explains why there's no data makes sense as a warning

File: parsers/CA_NS.py (hunk 2 above)
Comment (corradio) [reply parsers/CA_NS.py]: exactly. You can even drop the `CA-NS: ` prefix from the message.

**Comment by jarek** [comment]
Let's hope this works!

**Pull request merged**